some initial text for drand <> FVM FRC (#1)
drand/FIPs: FRCs/frc-00zz.md

## Simple Summary

> Timelock encryption is a method of encrypting a message now that can't be decrypted until a specific point in the future has passed. The drand team have launched a scheme that uses BLS12-381 signatures in an innovative way using [identity based encryption](some-link-here) to achieve practical Timelock encryption using the drand network.

Reviewer: ```suggestion
Timelock encryption is a method of encrypting a message now that can't be decrypted until a specific point in the future has passed. The drand team has launched a scheme that uses BLS12-381 signatures in an innovative way using [identity based encryption](some-link-here) to achieve practical Timelock encryption using the drand network.
```

Author: Both `have` and `has` are acceptable for collective nouns

> As the Filecoin network uses keypairs compatible with BLS12-381, it could trivially implement built-in FVM actors for enabling timelock encryption (providing the [FIP for switching to the new drand network](some-link-here) is accepted).

Reviewer: ```suggestion
As the Filecoin network already supports BLS12-381, it could trivially implement built-in FVM actors for enabling timelock encryption (providing the [FIP for switching to the new drand network](some-link-here) is accepted).
```

## Abstract

> On 2023-03-01, drand launched the first practical Timelock encryption scheme to mainnet after undergoing a security audit and developing libraries and tooling to make it easier.

Reviewer: ```suggestion
On 2023-03-01, drand launched the first practical Timelock encryption scheme to drand mainnet after undergoing a security audit and developing libraries and tooling to make it easier.
```

Timelock encryption would enable users to do sealed-bid auctions, create more secure implementations of [RANDAO](some-link-here), schedule social media network messages, prevent [MEV](some-link-here), and much, much more!
The drand timelock scheme uses [identity based encryption](some-link-here) which exploits the pairing capabilities of BLS12-381 to use a future round number as a public key to encrypt messages which will be decryptable only once that future round has been emitted by the drand network by using the round signature as a private key.

> We propose a precompile for FVM that would enable users to timelock encrypt arbitrary data and attempt decryption of previously timelock encrypted data. This would enable users to timelock encrypt some part of the data inside their smart contract, while also enabling users to pass ciphertexts directly into smart contracts for decryption there at a later time. 

Reviewer: Do we want some kind of auto-decryption cron job once time T is reached or really just "attempt decryption"?

Author: FVM are planning to remove cron jobs, so I don't know what the best solution is here; decryption oracles?

Reviewer 2: You can incentivizes ppl to submit these decryption onchain, or use something that ppl built already like https://www.gelato.network/

Author: I attended the IPC AMA yesterday, and I wonder if an incentivised timelock decryption subnet could be an option?! Could also be overkill but worth investigating

Reviewer 2: subnet I'm not sure because IPC <-> L1 communication may be expensive and take time but definitely an offchain loosely coordinated network, like Gelato. Or what Nois have been  doing with their randomness bots, everyone can submit drand randomness onchain and be compensated for that.

## Change Motivation
To enable timelock encryption and decryption using primitives that are already available in the FVM internals.

## Specification

The following interfaces should be fulfilled in FEVM and native actors respectively:

```solidity
interface Timelock {
    function encrypt(bytes message, uint64 roundNumber) returns (bytes)
    function decrypt(bytes message) returns (bytes, bool)

> }

Reviewer: How do we account for the fact the chainhash and public key of the network might change in the future? Should it be parameters too?

Author: Hmm good question - probably don't want to push this decision onto the caller, but perhaps deploy different precompiles for different networks. I'm not sure if there's prior art on FVM for this kind of decision

Author: I've added a line about this in the section

>     function encrypt(bytes message, uint64 roundNumber) returns (bytes)

Reviewer 2: I would not expose this API which is hard to use for users - I would either use "block number" from Filecoin or even better "timestamp" and then internally the API figure out which drand number it is.
Also this method in particular can not be done "onchain" so it's a bit weird to express it as a stg that "should be fulfilled in FEVM" , it has to be done on the client side.

Author: I think block number makes sense - timestamp could give the impression that everything will be exactly timely, and also that you could encrypt to rounds that won't be in the block headers of filecoin. Block number gives less precision, but it's the right precision for filecoin

Author: Have updated to that effect

```

```rust
trait Timelock {
    fn encrypt(message: Vec<u8>, roundNumber: u64) -> Result<Vec<u8>>
    fn decrypt(message: Vec<u8>) -> Result<Vec<u8>>
}
```

> The `encrypt` functions take raw message bytes and a round number which, along with the public key of the drand network, are passed into the function specified in [the timelock encryption paper](some-link-here) released by the team. In short, the round number is hashed to a point on the G2 extension field of the BLS12-381 curve. That point is multiplied by a new point derived from the message and mapped onto a target group on the Gt extension field, and the message is xor'd with the resulting point on the Gt extension field.

Reviewer: ```suggestion
The `encrypt` functions take raw message bytes and a round number which, along with the public key of the drand network, are passed into the function specified in [the timelock encryption paper](some-link-here) released by the team. In short, the round number is hashed to a point on the G2 group of the BLS12-381 curve. That point is multiplied by a new point derived from the message and mapped onto the target group Gt, and the message is xor'd with the resulting point on Gt.
```

> For larger messages originating outside the system, we recommend using a symmetric cipher such as [AGE](some-link-here) to encrypt the message, and performing timelock encryption on just the symmetric key, in order to save gas fees, as the pairing operation is comparatively expensive.

Reviewer: ```suggestion
For larger messages originating outside the system, we recommend using a symmetric cipher such as [age](https://age-encryption.org/) to encrypt the message, and performing timelock encryption on just the symmetric key, in order to save gas fees, as the pairing operation is comparatively expensive.
```

Is doing on-chain `age` decryption realistic?

Reviewer: Or should we consider having a XOF based system doing just XOR?

Author: >Is doing on-chain age decryption realistic?
I imagine it's _technically_ possible, but would be very expensive. 
+ cheaper gas fees
+ more flexible if users wish to use other schemes
- we'd need to modify the tlock libraries to support this scenario
- it gives users more scope to do it wrong and try and encrypt huge payloads with tlock

Reviewer: I don't think age on-chain has cheaper gas fees! On the contrary

Reviewer: So is the message here that users shouldn't timelock their content onchain but only their keys? That could be clearer I think

Author: oops yeah dunno why I said it has cheaper gas fees? also the first to were meant to be + and the latter two -

Agree re: messaging, will update

edit: ah I meant using AGE off-chain and only encrypting the symmetric key. Users will have to _know_ to do the symmetric encryption off-chain, and there's a risk that they'll still try and encrypt a huge payload with timelock on-chain

We propose a precompile for FVM that would enable users to timelock encrypt arbitrary data and attempt decryption of previously timelock encrypted data. This would enable users to timelock encrypt some part of the data inside their smart contract, while also enabling users to pass ciphertexts directly into smart contracts for decryption there at a later time. 

## Change Motivation
To enable timelock encryption and decryption using primitives that are already available in the FVM internals.

## Specification

The following interfaces should be fulfilled in FEVM and native actors respectively:

```solidity
interface Timelock {
    function encrypt(bytes message, uint64 blockNumber) returns (bytes)
    function decrypt(bytes message) returns (bytes, bool)
}
```

```rust
trait Timelock {
    fn encrypt(message: Vec<u8>, blockNumber: u64) -> Result<Vec<u8>>
    fn decrypt(message: Vec<u8>) -> Result<Vec<u8>>

> }

Reviewer 2: I would separate exactly what should be available in the FEVM (decrypt) and what should be implemented by wallets/clients (encrypt).

Author: I envisaged that users could also encrypt from contract to contract - of course this gives the miners some advantage but there still could be valid uses cases.
A related issue is that if users do encryption off-chain, they will have to ensure that they use a round that is included in the filecoin headers; we can write a separate client to make that seamless though

Reviewer 2: But what does it mean encrypting from contract ? The plaintext is public, for everyone, not just miners, since everything that gets executed in the FEVM is public.

Reviewer: That's a fair point xD It's indeed more interesting to be able to decrypt data than to encrypt data

Author: mmm yeah I didn't think that through very well... it makes no sense at all to provide encrypt capabilities

Author: In that case maybe the symmetric encryption discussion below is also kind of moot - unless the wider tooling provides capabilities for doing that easily?

Author: Well not totally, symmetric decryption is still a debate

Reviewer: yup decryption needs to happen

```

The `encrypt` functions take raw message bytes and a block number which, along with the public key of the drand network, are passed into the function specified in [the timelock encryption paper](https://eprint.iacr.org/2023/189.pdf) released by the team. In short, the block number is hashed to a point on the G2 group of the BLS12-381 curve. That point is multiplied by a new point derived from the message and mapped onto the target group Gt, and the message is xor'd with the resulting point on Gt.

> For larger messages originating outside the system, we recommend using a symmetric cipher such as [age](https://age-encryption.org/) to encrypt the message off-chain, and performing timelock encryption on just the symmetric key on-chain, in order to save gas fees, as the pairing operation is expensive.

Reviewer 2: If that's the recommendation of the FIP then it must come with the spec or stg with the FIP ? 
I mean the FIP should be self contained, so looking at it we should be able to answer the question "how do we encrypt the transaction" "how do we decrypt it". 
I would recommend for a very simple format, but given age is well documented maybe that can fly. The only thing that bothers me is that it's not doable in Solidity/EVM land to decrypt age so maybe it's worth finding a AEAD very simple that is implementable in solidity land.  (but it's up for discussion to devs I guess if they're ok implementing the "decrypt" as a precompile or not)

Author: Perhaps it's worth having another FRC for best practices around AEAD on FVM/FEVM? 
I had a look for some prior art on ethereum, but the best I could find was [this draft proposal](https://eips.ethereum.org/EIPS/eip-5630) which quite loosely says "use ECIES + AES-CBC", but doesn't really provide any implementation details.

I don't see any impls of common symmetric ciphers on github in solidity, but I think we could port our impl of Chacha20Poly1305 to solidity fairly easily; whether that would be peformant or cheap... I'm not so sure.
A benefit of raising a separate FRC for it would be that it could be another rust precompile and comparatively cheaper

## Simple Summary

> Timelock encryption is a method of encrypting a message now that can't be decrypted until a specific point in the future has passed. The drand team have launched a scheme that uses BLS12-381 signatures in an innovative way using [identity based encryption](https://crypto.stanford.edu/~dabo/papers/bfibe.pdf) to achieve practical Timelock encryption using the drand network.

Reviewer: Reference the tlock paper maybe?

https://eprint.iacr.org/2023/189

Author: done

> As the Filecoin network already supports BLS12-381, it could trivially implement built-in FVM actors for enabling timelock encryption (providing the [FIP for switching to the new drand network](../FIPS/fip-0063.md) is accepted).

Reviewer: ```suggestion
As the Filecoin network already supports BLS12-381, it could implement built-in FVM actors for enabling timelock encryption (providing the [FIP for switching to the new drand network](../FIPS/fip-0063.md) is accepted).
```

## Abstract

> On 2023-03-01, drand launched the first practical Timelock encryption scheme to drand mainnet after undergoing a security audit and developing libraries and tooling to make it easier.

Reviewer: Should we hold on to see if we're re-spawning a new `fastnet` network with the patched signatures during the ceremony on June 7th or 14th?

Author: yeah, let's wait

> Timelock encryption would enable users to do sealed-bid auctions, create more secure implementations of [RANDAO](https://github.com/randao/randao), schedule social media network messages, prevent [MEV](https://coinmarketcap.com/alexandria/glossary/miner-extractable-value-mev), and much, much more!

Reviewer: Sealed NFT airdrops come to mind too.

Author: added
